### PR TITLE
[data][doc] auto-generate dataset api documentation

### PR DESCRIPTION
**Pull request by can-anyscale**
This PR auto-generate dataset API documentation, by moving the declaration of API group to the code itself. Also few other fixes to bring consistency between pages.

Test:
- dataset page: https://anyscale-ray--46557.com.readthedocs.build/en/46557/data/api/dataset.html
- navigation page: https://anyscale-ray--46557.com.readthedocs.build/en/46557/data/api/api.html

Note that the  documentation is now consistently documenting a class construction in its full form

### CODE REVIEW

File: doc/source/_templates/autosummary/class_v2.rst [hunk 1]
```diff
@@ -1,4 +1,5 @@
-{{ name | escape | underline}}
+{{ name }}
+{{ '-' * name | length }}
```
Comment (can-anyscale) [inline doc/source/_templates/autosummary/class_v2.rst:2]: change to `H2` instead of `H1` header type

File: doc/source/conf.py [hunk 2]
```diff
@@ -401,7 +401,10 @@ def get_api_groups(method_names, class_name, module_name):
     cls = getattr(import_module(module_name), class_name)
     for method_name in method_names:
         method = getattr(cls, method_name)
-        api_groups.add(safe_getattr(method, "_annotated_api_group", DEFAULT_API_GROUP))
+        if _is_public_api(method):
```
Comment (can-anyscale) [inline doc/source/conf.py:404]: only need to get api group from public APIs

File: doc/source/data/api/data_iterator.rst [hunk 3]
```diff
@@ -1,3 +1,6 @@
 .. _dataset-iterator-api:
 
+DataIterator API
```
Comment (can-anyscale) [inline doc/source/data/api/data_iterator.rst:3]: to be consistent with how dataset api is documented

File: python/ray/util/annotations.py [hunk 4]
```diff
@@ -46,16 +46,14 @@ def PublicAPI(*args, **kwargs):
         ...     return y
     """
     if len(args) == 1 and len(kwargs) == 0 and callable(args[0]):
-        return PublicAPI(stability="stable", api_group="others")(args[0])
+        return PublicAPI(stability="stable", api_group="Others")(args[0])
 
     if "stability" in kwargs:
         stability = kwargs["stability"]
         assert stability in ["stable", "beta", "alpha"], stability
-    elif kwargs:
```
Comment (can-anyscale) [inline python/ray/util/annotations.py:54]: remove this condition since the key can also be `api_group` (in addition to `stability`)

**Review by bveeramani — APPROVED** [review]
Nice!!

**Pull request merged**